common/src/test/java/org/conscrypt/javax/crypto/CipherBasicsTest.java

```diff
@@ -240,6 +251,47 @@ public void testAeadEncryption() throws Exception {
         }
     }
 
+    public void bufferBasedAssessment(Cipher cipher, byte[] aad, byte[] tag, byte[] _plaintext,
+                                           byte[] _ciphertext, Key key, AlgorithmParameterSpec params,
+                                           String transformation, Provider p) throws Exception {
+        cipher.init(Cipher.ENCRYPT_MODE, key, params);
+        if (aad.length > 0) {
+            cipher.updateAAD(aad);
+        }
+        byte[] _combinedOutput = new byte[_ciphertext.length + tag.length];
```

Comment:
Looking good so far, but we should test all the combinations here: 4 obvious ones for plainText and outputBuffer being direct / non-direct and one final version where they are both non-direct and share the same backing array. Otherwise I don't think we're covering all the new code.

Reply:
I added some boolean gates and then called all 4 combinations. I created a separate function for the shared backing as well since I felt it would be too complicated to pick and choose with the other combos